src/util/merkleTree.ts

```diff
@@ -0,0 +1,150 @@
+/**
+ * @module util/merkleTree
+ */
+
+import assert from "assert";
+import {bytes32} from "../types";
+import {hash} from "../util/crypto";
+import {intDiv} from "../util/math";
+
+interface IProgressiveMerkleTree {
+  /**
+   * The number of items in the tree
+   */
+  count(): number;
+  depth(): number;
+  /**
+   * Add a new item into the tree
+   *
+   * Return a proof to verify the item is in the tree
+   * @returns proof
+   */
+  push(item: bytes32): bytes32[];
+  /**
+   * The merkle root of the tree
+   */
+  root(): bytes32;
+}
+
+export class ProgressiveMerkleTree implements IProgressiveMerkleTree {
+  private _depth: number;
+  private _count: number;
+  private _branch: bytes32[];
+  private _zerohashes: bytes32[];
+
+  public constructor(depth: number) {
+    assert(depth <= 52, "tree depth must be less than 53");
```

Comment:
Couldn't find reference, why 52? :)

Reply:
It's so we don't overflow js numbers. There's a part in the code where we may try to calculate 2**depth.